pkg/kubelet/cm/node_allocatable.go

```diff
@@ -0,0 +1,164 @@
+	systemReservedEnforcementKey     = "system-reserved"
+	kubeReservedEnforcementKey       = "kube-reserved"
+)
+
+func createNodeAllocatableCgroups(nc NodeAllocatableConfig, nodeAllocatable v1.ResourceList, cgroupManager CgroupManager) error {
+	cgroupConfig := &CgroupConfig{
+		Name: CgroupName(defaultNodeAllocatableCgroupName),
+	}
+	if err := cgroupManager.Create(cgroupConfig); err != nil {
+		glog.Errorf("Failed to create %q cgroup and apply limits")
+		return err
+	}
+	return nil
+}
+
+// Enforce Node Allocatable Cgroup settings.
+func enforceNodeAllocatableCgroups(nc NodeAllocatableConfig, nodeAllocatable v1.ResourceList, cgroupManager CgroupManager) error {
+	glog.V(4).Infof("Attempting to enforce Node Allocatable with config: %+v", nc)
+	glog.V(4).Infof("Node Allocatable resources: %+v", nodeAllocatable)
+	// Create top level cgroups for all pods if necessary.
+	if nc.EnforceNodeAllocatable.Has(nodeAllocatableEnforcementKey) {
+		cgroupConfig := &CgroupConfig{
+			Name:               CgroupName(defaultNodeAllocatableCgroupName),
+			ResourceParameters: getCgroupConfig(nodeAllocatable),
+		}
+		glog.V(4).Infof("Updating Node Allocatable cgroup with %d cpu shares and %d bytes of memory", cgroupConfig.ResourceParameters.CpuShares, cgroupConfig.ResourceParameters.Memory)
+		if err := cgroupManager.Update(cgroupConfig); err != nil {
+			glog.Errorf("Failed to create %q cgroup and apply limits")
+			return err
+		}
+	}
+	// Now apply kube reserved and system reserved limits if required.
+	if nc.EnforceNodeAllocatable.Has(systemReservedEnforcementKey) {
+		glog.V(2).Infof("Enforcing system reserved on cgroup %q with limits: %+v", nc.SystemReservedCgroupName, nc.SystemReserved)
+		if err := enforceExistingCgroup(cgroupManager, nc.SystemReservedCgroupName, nc.SystemReserved); err != nil {
+			return fmt.Errorf("failed to enforce System Reserved Cgroup Limits: %v", err)
+		}
+	}
+	if nc.EnforceNodeAllocatable.Has(kubeReservedEnforcementKey) {
+		glog.V(2).Infof("Enforcing kube reserved on cgroup %q with limits: %+v", nc.KubeReservedCgroupName, nc.KubeReserved)
+		if err := enforceExistingCgroup(cgroupManager, nc.KubeReservedCgroupName, nc.KubeReserved); err != nil {
+			return fmt.Errorf("failed to enforce Kube Reserved Cgroup Limits: %v", err)
+		}
+	}
+	return nil
+}
+
+func enforceExistingCgroup(cgroupManager CgroupManager, cName string, rl v1.ResourceList) error {
+	cgroupConfig := &CgroupConfig{
+		Name:               CgroupName(cName),
+		ResourceParameters: getCgroupConfig(rl),
+	}
+	glog.V(4).Infof("Enforcing limits on cgroup %q with %d cpu shares and %d bytes of memory", cName, cgroupConfig.ResourceParameters.CpuShares, cgroupConfig.ResourceParameters.Memory)
+	if !cgroupManager.Exists(cgroupConfig.Name) {
+		return fmt.Errorf("%q cgroup does not exist", cgroupConfig.Name)
+	}
+	if err := cgroupManager.Update(cgroupConfig); err != nil {
+		return err
+	}
+	return nil
+}
+
+func getCgroupConfig(rl v1.ResourceList) *ResourceConfig {
```

Comment:
nit: i think this should make clear its a cgroup config for guaranteed QoS since there could be another method laying around that gets a cgroup config that takes a resource requirements spec.